src/Features/Core/Portable/Snippets/SnippetProviders/AbstractForLoopSnippetProvider.cs

```diff
@@ -0,0 +1,102 @@
+        {
+            var snippetTextChange = await GenerateSnippetTextChangeAsync(document, position, cancellationToken).ConfigureAwait(false);
+            return ImmutableArray.Create(snippetTextChange);
+        }
+
+        private async Task<TextChange> GenerateSnippetTextChangeAsync(Document document, int position, CancellationToken cancellationToken)
+        {
+            var forLoopSyntax = await CreateForLoopStatementSyntaxAsync(document, cancellationToken).ConfigureAwait(false);
+            return new TextChange(TextSpan.FromBounds(position, position), forLoopSyntax.NormalizeWhitespace().ToFullString());
+        }
+
+        /// <summary>
+        /// Tries to get the location after the open parentheses in the argument list.
+        /// If it can't, then we default to the end of the snippet's span.
+        /// </summary>
+        protected override int GetTargetCaretPosition(ISyntaxFactsService syntaxFacts, SyntaxNode caretTarget)
+        {
+            return GetCaretPosition(caretTarget);
+        }
+
+        protected override async Task<SyntaxNode> AnnotateNodesToReformatAsync(Document document,
+            SyntaxAnnotation findSnippetAnnotation, SyntaxAnnotation cursorAnnotation, int position, CancellationToken cancellationToken)
+        {
+            var root = await document.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+            var syntaxFacts = document.GetRequiredLanguageService<ISyntaxFactsService>();
+            var snippetExpressionNode = FindAddedSnippetSyntaxNode(root, position, syntaxFacts);
+            if (snippetExpressionNode is null)
+            {
+                return root;
+            }
+
+            var reformatSnippetNode = snippetExpressionNode.WithAdditionalAnnotations(findSnippetAnnotation, cursorAnnotation, Simplifier.Annotation, Formatter.Annotation);
+            return root.ReplaceNode(snippetExpressionNode, reformatSnippetNode);
+        }
+
+        protected override ImmutableArray<SnippetPlaceholder> GetPlaceHolderLocationsList(SyntaxNode node, ISyntaxFacts syntaxFacts, CancellationToken cancellationToken)
+        {
+            return GetForLoopSnippetPlaceholders(node, syntaxFacts);
+        }
+
+        protected override SyntaxNode? FindAddedSnippetSyntaxNode(SyntaxNode root, int position, ISyntaxFacts syntaxFacts)
+        {
+            var closestNode = root.FindNode(TextSpan.FromBounds(position, position), getInnermostNodeForTie: true);
+            var nearestStatement = closestNode.DescendantNodesAndSelf(syntaxFacts.IsForStatement).FirstOrDefault();
+            if (nearestStatement is null)
+            {
+                return null;
+            }
+
+            // Checking to see if that expression statement that we found is
+            // starting at the same position as the position we inserted
+            // the for statement.
+            if (nearestStatement.SpanStart != position)
```

Comment:
if you have the starting position, and you're checking that they must match ehre, then there's no need to do DescendentNodes afaict...